addons/mail/controllers/mail.py

```diff
@@ -197,3 +201,39 @@ def mail_action_unfollow(self, model, res_id, pid, token, **kwargs):
             'model_name': request.env['ir.model'].sudo()._get(model).display_name,
             'access_url': record._notify_get_action_link('view', model=model, res_id=res_id) if display_link else False,
         })
+
+    @http.route('/mail/message/<int:message_id>', type='http', auth='public')
+    @add_guest_to_context
+    def mail_thread_message_redirect(self, message_id, **kwargs):
+        message = request.env['mail.message'].search([('id', '=', message_id)])
+        if not message:
+            if request.env.user._is_public():
+                return request.redirect(f'/web/login?redirect=/mail/message/{message_id}')
+            raise Unauthorized()
+
+        # sudo: public user can access some relational fields of mail.message
+        if message.sudo()._filter_empty():
+            raise NotFound()
+        if not request.env.user._is_internal():
+            thread = request.env[message.model].search([('id', '=', message.res_id)])
+            if message.model == 'discuss.channel':
+                store = Store({'isChannelTokenSecret': True})
+                store.add(thread, {'highlightMessage': Store.one(message, only_id=True)})
+                return PublicPageController()._response_discuss_channel_invitation(store, thread)
+            else:
+                return request.redirect(thread._get_share_url(share_token=False))
```

Comment:
How did you pass any testing phase ? This method is defined in portal and defined only on portal.mixin inheriting models ...